ecp5/gfx.cc

```diff
@@ -0,0 +1,1484 @@
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "gfx.h"
+
+NEXTPNR_NAMESPACE_BEGIN
+
+void gfxTileBel(std::vector<GraphicElement> &g, int x, int y, int z, int w, int h, IdString bel_type,
+                GraphicElement::style_t style)
+{
+    GraphicElement el;
+    el.type = GraphicElement::TYPE_BOX;
+    el.style = style;
+    if (bel_type == id_TRELLIS_SLICE) {
+        el.x1 = x + slice_x1;
+        el.x2 = x + slice_x2;
+        el.y1 = y + slice_y1 + (z)*slice_pitch;
+        el.y2 = y + slice_y2 + (z)*slice_pitch;
+        g.push_back(el);
+
+        el.style = GraphicElement::STYLE_FRAME;
+        el.x1 = x + slice_x2 + 0.0255f;
+        el.x2 = el.x1 + 0.0017f;
+        el.y1 = y + slice_y2 - 0.0017f * (TILE_WIRE_CLK3_SLICE - TILE_WIRE_DUMMY_D2 + 5 + (3 - z) * 26) +
+                3 * slice_pitch - 0.0007f;
+        el.y2 = el.y1 + 0.0017f * 5;
+        g.push_back(el);
+    } else if (bel_type == id_TRELLIS_IO || bel_type == id_IOLOGIC || bel_type == id_SIOLOGIC ||
+               bel_type == id_DQSBUFM) {
+        bool top_bottom = (y == 0 || y == (h - 1));
+        if (top_bottom) {
+            el.x1 = x + io_cell_h_x1 + (z + 2) * 0.10;
+            el.x2 = x + io_cell_h_x1 + (z + 2) * 0.10 + 0.08f;
+            if (y == h - 1) {
+                el.y1 = y + 1 - io_cell_h_y1;
+                el.y2 = y + 1 - io_cell_h_y2;
+            } else {
+                el.y1 = y + io_cell_h_y1;
+                el.y2 = y + io_cell_h_y2;
+            }
+        } else {
+            if (x == 0) {
+                el.x1 = x + 1 - io_cell_v_x1;
+                el.x2 = x + 1 - io_cell_v_x2;
+            } else {
+                el.x1 = x + io_cell_v_x1;
+                el.x2 = x + io_cell_v_x2;
+            }
+            el.y1 = y + io_cell_v_y1 + z * 0.10;
+            el.y2 = y + io_cell_v_y1 + z * 0.10 + 0.08f;
+        }
+        g.push_back(el);
+    } else if (bel_type == id_DCCA) {
+        el.x1 = x + switchbox_x1 + (z)*0.025;
+        el.y1 = y + 0.14;
+        el.x2 = x + switchbox_x1 + (z)*0.025 + 0.020;
+        el.y2 = y + 0.18;
+        g.push_back(el);
+    } else if (bel_type == id_DP16KD || bel_type == id_MULT18X18D || bel_type == id_ALU54B) {
+        el.x1 = x + slice_x1;
+        el.x2 = x + 0.97;
```

Comment:
imo, frequently used numbers for GUI purposes should be constants so GUI layout changes are easier